### PR TITLE
[feature](Nereids): add rule split limit into two phase

### PR DESCRIPTION
**Pull request by keanji-x**
# Proposed changes
* Add a rule split limit, like Limit(Origin) ==> Limit(Global) -> Gather -> Limit(Local)
* Add a rule: limit-> sort ====> topN
* fix a bug about topN
* make the type of limit,offset long in topN

clos #16693

And because this rule is always beneficial, we add a rule in the rewrite phase
## Problem summary

Describe your changes.

## Checklist(Required)

* [ ] Does it affect the original behavior
* [x] Has unit tests been added
* [ ] Has document been added or modified
* [ ] Does it need to update dependencies
* [ ] Is this PR support rollback (If NO, please explain WHY)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 35.15 seconds
 stream load tsv:          477 seconds loaded 74807831229 Bytes, about 149 MB/s
 stream load json:         36 seconds loaded 2358488459 Bytes, about 62 MB/s
 stream load orc:          69 seconds loaded 1101869774 Bytes, about 15 MB/s
 stream load parquet:          28 seconds loaded 861443392 Bytes, about 29 MB/s
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230215141917_clickbench_pr_97184.html

File: fe/fe-core/src/main/java/org/apache/doris/nereids/glue/translator/PhysicalPlanTranslator.java [hunk 1]
```diff
@@ -1338,38 +1338,15 @@ public PlanFragment visitPhysicalLimit(PhysicalLimit<? extends Plan> physicalLim
...
-        PlanFragment planFragment = exchangeToMergeFragment(inputFragment, context);
-        planFragment.getPlanRoot().setLimit(physicalLimit.getLimit());
-        planFragment.getPlanRoot().setOffSetDirectly(physicalLimit.getOffset());
-        return planFragment;
+        child.setLimit(physicalLimit.getLimit());
+        child.setOffset(physicalLimit.getOffset());
+        // TODO: add a rule in cascades
+        //      before: GlobalLimit -> LocalLimit -> GlobalSort -> Gather -> LocalSort
+        //      after: GlobalLimit -> GlobalSort -> Gather ->LocalLimit -> LocalSort
+        return inputFragment;
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/glue/translator/PhysicalPlanTranslator.java:1349]: if we have not rule to rewrite Limit(Sort) to TopN, and do not set limit and offset on ExchangeNode's child, the performance of `order by x limit n` is not acceptable.

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/logical/LogicalLimit.java [hunk 2]
```diff
@@ -44,19 +45,28 @@
...
 public class LogicalLimit<CHILD_TYPE extends Plan> extends LogicalUnary<CHILD_TYPE> implements Limit {
-
+    private final LimitPhase phase;
     private final long limit;
     private final long offset;
 
-    public LogicalLimit(long limit, long offset, CHILD_TYPE child) {
-        this(limit, offset, Optional.empty(), Optional.empty(), child);
+    public LogicalLimit(long limit, long offset, CHILD_TYPE child, LimitPhase phase) {
+        this(limit, offset, Optional.empty(), Optional.empty(), child, phase);
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/logical/LogicalLimit.java:53]: children should be the last parameter

Comment (XieJiann) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/logical/LogicalLimit.java]: fixed

File: fe/fe-core/src/main/java/org/apache/doris/nereids/glue/translator/PhysicalPlanTranslator.java (hunk 1 above)
Comment (XieJiann) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/glue/translator/PhysicalPlanTranslator.java]: add a rule limit->sort => topn in LimitPushDown Rule

**Comment by XieJiann** [comment]
run buildall

**Comment by XieJiann** [comment]
run buildall

**Comment by XieJiann** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/physical/PhysicalLimit.java [hunk 3]
```diff
@@ -70,14 +72,16 @@ public PhysicalLimit(long limit, long offset,
      *
      * @param limit the number of tuples retrieved.
      * @param offset the number of tuples skipped.
+     * @param phase the phase of 2-phase limit.
      */
     public PhysicalLimit(long limit, long offset, Optional<GroupExpression> groupExpression,
             LogicalProperties logicalProperties, PhysicalProperties physicalProperties,
-            StatsDeriveResult statsDeriveResult, CHILD_TYPE child) {
+            StatsDeriveResult statsDeriveResult, CHILD_TYPE child, LimitPhase phase) {
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/physical/PhysicalLimit.java:79]: do not put phase in the last

File: fe/fe-core/src/test/java/org/apache/doris/rewrite/SplitLimit.java [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+import org.apache.doris.catalog.PrimitiveType;
+import org.apache.doris.nereids.trees.plans.Plan;
+import org.apache.doris.nereids.trees.plans.logical.LogicalLimit;
+import org.apache.doris.nereids.util.PlanChecker;
+import org.apache.doris.utframe.TestWithFeService;
+
+import org.junit.jupiter.api.Test;
+
+public class SplitLimit extends TestWithFeService {
```
Comment (morrySnow) [inline fe/fe-core/src/test/java/org/apache/doris/rewrite/SplitLimit.java:28]: start a mock service is very heavy, so it is better to use PlanChecker to do rule UT

**Comment by XieJiann** [comment]
run buildall

**Comment by XieJiann** [comment]
run buildall

**Review by morrySnow — APPROVED** [review]

**Pull request merged**